docs/sdk-spec.markdown

```diff
@@ -0,0 +1,256 @@
+types of flashing modes.
+
+API
+---
+
+The Etcher SDK will expose this API:
+
+### `String[] etcher.SUPPORTED_MODES`
+
+An array of the modes this particular version of the Etcher SDK supports.
+
+### `EventEmitter etcher.image.fromURL(String url)`
+
+This function creates an *image* object out of an image URL. The implementation
+will at least support HTTP, HTTPS, and `file://`.
+
+- `String url`: The url to the image file
+
+The returned `EventEmitter` emits these events:
+
+- `done`: The operation was successful, and the handler argument is the image
+  object
+- `progress` (optional): The progress details, useful when constructing an
+  image object for a certain image will take more time
+- `error`: An error occurred
+
+### `EventEmitter etcher.image.fromFile(String path)`
+
+This function is like calling `etcher.image.fromURL()` with a `file://`
+protocol.
+
+### `String[] etcher.getSupportedModesForImage(Image image)`
+
+This function returns a list of supported modes for a particular image object.
+
+### `Promise<Destination[]> etcher.scan(Mode mode, Object options)`
+
+This function scans for destinations, passing the `options` object to the
+scanner's `.getAll()` function as-is.
+
+If the scanner singleton is not running, this function will yield back an empty
+set of destinations.
+
+### `EventEmitter etcher.flash(Mode mode, Image image, Destination[] destinations, Object options)`
+
+This function flashes an image to a set of destinations using a certain mode.
+
+The options object may contain these properties, although there is no guarantee
+that every mode will support them all.
+
+- `Boolean check`: Check the result of a flashing process, in whatever way
+  makes sense on that mode
+- `Object extra`: Extra options specific to the chosen mode
+
+The returned `EventEmitter` emits these events:
+
+- `done`: The operation was successful, and the handler argument is a mode
+  specific results object
+- `progress`: Flashing progress details
+- `error`: An error occurred
+
+The flash implementation may emit custom events, as long as they include an
+'extra:' prefix.
+
+Interfaces
+----------
+
+The SDK deals with various "entities." This is a formal definition of what
+these entities look like:
+
+### Progress
+
+An in-progress action. A progress object may include these properties:
+
+- `String id`: A unique identifier for what the progress is about
+- `Number percentage`: The progress percentage
+- `Number transferred`: The number of transferred parts (whatever that means in
+  the progress context)
+- `Number length`: The total number of transferred parts
+- `Number remaining`: The number of parts that are pending
+- `Number eta`: The number of estimated remaining seconds
+- `Number runtime`: The number of seconds the progress was active
+- `Number speed`: The average number of parts processed in each tick
+
+### Image
+
+An operating system image file, like `2016-11-25-raspbian-jessie.img`, or
```

Comment:
this image could be a zip, elf or bin file. how would we encode the extra properties into these img types?